### ISSUE TITLE
Can't create a new file after first run as root

### ISSUE THREAD

**Issue by Janiczek** [elementary/code#202]
When I try to create a new file from the main window ("new file / open file / new from clipboard"), the app doesn't initialize the text editor window and starts looking like this (still responsive, but I can't write):

![from_main_window](https://user-images.githubusercontent.com/149425/29780760-3829a864-8c17-11e7-9801-ac97047681c1.png)

When I try to create a new file from an opened file (by clicking on the `+` sign in the tab list), similar story (but I can still edit the previously open file) - the plus sign stays "focused", but nothing else happens:

![plus](https://user-images.githubusercontent.com/149425/29780890-b18ee764-8c17-11e7-8040-7c01897cf7f1.png)

**Comment by Janiczek** [comment]
I have built `io.elementary.code` from source and run it, logging ot STDOUT. It gave this error when clicking on the "New File" button:

```
[FATAL 22:43:59.172485] DocumentView.vala:120: Error opening file '/home/martin/.local/share/scratch-text-editor/unsaved/Text file from 2017-08-28 22:43:59': Permission denied
```

The directory structure has these permissions:
```
rwx------ ~
rwx------ ~/.local
rwx------ ~/.local/share
rwxr-xr-x ~/.local/share/scratch-text-editor
rwxr-xr-x ~/.local/share/scratch-text-editor/unsaved
```

(This is from clean `elementary OS` installation.)

When I added the write permissions (`chmod 777`) to `scratch-text-editor` and `unsaved` directories, **it started working.**

I don't understand the cause - maybe some installation script?

**Comment by matzipan** [comment]
Does "clean" elementary OS install mean you did not copy the home directory from another installation?

**Comment by matzipan** [comment]
Can you also do a ls -al /home/martin/.local/share? Perhaps first time when you ran Code it was run as root or something and the directory in there was created as root, which is why the permissions did not allow you to write to it.

**Comment by Janiczek** [comment]
Correct, I have not copied the home directory from another installation. (Only post-installation I copied some data like photos etc., but no dotfiles / directories.)

```
martin@janiczek:~$ ls -al /home/martin/.local/share
total 176
drwxr-xr-x 25 martin martin  4096 srp 29 09:25 .
drwxr-xr-x  3 martin martin  4096 srp 25 01:37 ..
drwx------  2 martin martin  4096 srp 29 09:11 applications
drwx------  2 martin martin  4096 srp 25 09:28 desktop-directories
drwx------  4 martin martin  4096 srp 24 23:38 epiphany
drwx------  7 martin martin  4096 srp 25 01:37 evolution
drwxr-xr-x  2 martin martin  4096 srp 25 01:37 gnome-settings-daemon
-rw-rw-r--  1 martin martin   505 srp 25 10:14 gsettings-data-convert
drwx------  2 martin martin  4096 srp 29 09:11 gvfs-metadata
drwxrwxr-x  2 martin martin  4096 srp 25 14:41 icc
drwx------  3 martin martin  4096 srp 25 09:26 icons
drwxrwxr-x  3 martin martin  4096 srp 25 10:43 JetBrains
drwxr-xr-x  6 martin martin  4096 srp 25 11:42 keybase
drwx------  2 martin martin  4096 srp 29 09:38 keyrings
drwx------  2 martin martin  4096 srp 25 11:53 libgda
drwxrwxr-x  3 martin martin  4096 srp 25 14:16 messenger
drwxrwxr-x  2 martin martin  4096 srp 25 18:17 nasc
drwxrwxr-x  2 martin martin  4096 srp 25 11:53 noise
drwx------  5 martin martin  4096 srp 29 09:23 nvim
drwxrwxr-x  2 martin martin  4096 srp 25 13:43 pantheon-photos
drwxrwxr-x  3 martin martin  4096 srp 25 01:37 plank
-rw-------  1 martin martin 20480 srp 29 09:25 recently-used.xbel
-rw-rw-r--  1 martin martin 13357 srp 28 11:36 recently-used.xbel.KKUX5Y
-rw-rw-r--  1 martin martin  2356 srp 25 10:54 recently-used.xbel.WT5D5Y
drwxrwxrwx  3 root   root    4096 srp 24 23:40 scratch-text-editor
drwx------  2 martin martin  4096 srp 25 01:37 sounds
drwx------  5 martin martin  4096 srp 25 15:39 Trash
drwx------  3 martin martin  4096 srp 25 13:49 webkitgtk
drwx------  3 martin martin  4096 srp 25 01:37 zeitgeist
```

I have also started on 0.4.1 Loki but then added the `daily` and `os-patches` PPAs and upgraded everything, but this issue was present before that change.

**Comment by matzipan** [comment]
Well, as you can see on the `scratch-text-editor` line, it says the folder is owned by root and belongs to the root group, which means when you first ran Code, you ran it as root (probably with `sudo`). So `chmod 755 scratch-text-editor && sudo chown martin:martin scratch-text-editor` should fix your issue.

**Comment by Janiczek** [comment]
Is it a good idea to show some kind of message to the user? ("Can't open a new file, click this button to fix this - might want sudo permissions" or something similar)

**Comment by matzipan** [comment]
It's a bug caused by an edge case. We definetely can't cover all posibilities. I consider that the error message "Permission denied" is descriptive enough.

**Comment by donadigo** [comment]
If the saved file was saved as root, and it cannot be read by Scratch, it should just ignore it and create a new one. Also: even if Scratch is opened as root, the file should be saved with right permissions and not with root permissions.

**Comment by matzipan** [comment]
Well I don't think it's because the file was saved as root, it's that the directory was created as root so you can't create a new file in there.

**Comment by Janiczek** [comment]
Just want to add that the error Permission denied can't be seen anywhere in the app. With some developer effort the user can run the app from terminal and see it, but still...

**Comment by matzipan** [comment]
You don't need to compile the app from scratch to run from the command line, just run `io.elementary.code` in the terminal.

**Comment by Janiczek** [comment]
Yes, I didn't mean to say the user has to compile the app (although I did). But there still probably will be some developer guidance needed to tell the user to run the app in the terminal. (Seeing as the elementary OS targets non-tech-savvy people too, I think a hint in the app could help, although that's up for discussion.)

**Comment by matzipan** [comment]
You're right, it should handle this more gracefully.

**Comment by danrabbit** [comment]
The probably solution is probably to completely disallow running with sudo as we’ve done in files and only allow running with pkexec

**Comment by danrabbit** [comment]
Relevant files commit here: https://github.com/elementary/files/commit/f0e4a4471e37c93ad692cbe376ce86e8d8e8a4a3

**Comment by linkert** [comment]
> The proper solution is probably to completely disallow running with sudo as we’ve done in files and only allow running with pkexec so that this can’t happen again

Sorry to quote reply such an old comment but I found it relevant as I filed issue #546. How do I modify locked files then if the proper solution is to have Code be unable to run under sudo all together? It's something that people do from time to time on Linux, BSD and macOS right? To me, as the end user I would much rather use Code over whipping out nano, nvim, ed or any other cli editor for editing locked files - for whatever reason I'm editing the files.
